File: hackerrank/problem solving/algorithms/2 - implementation/32 - sherlock and squares/with morfo library/sherlock_and_squares.py
# https://www.hackerrank.com/challenges/sherlock-and-squares/problem?isFullScreen=true

import math

import morfo


def main():
    n = morfo.read(int)
    for _ in range(n):
        a = morfo.read(int)
        b = morfo.read(int)
        print(squares(a, b))


# T: O(1)
# S: O(1) extra space
def squares(start_num, end_num):
    max_square = int(math.floor(math.sqrt(end_num)))
    min_square = int(math.ceil(math.sqrt(start_num)))
    return max_square - min_square + 1


if __name__ == '__main__':
    main()

File: hackerrank/problem solving/algorithms/2 - implementation/32 - sherlock and squares/with morfo library/sherlock_and_squares.go
// https://www.hackerrank.com/challenges/sherlock-and-squares/problem?isFullScreen=true

package main

import (
	"fmt"
	"math"
	"morfo/morfoio"
)

func main() {
	n := morfoio.Read[int]()
	for range n {
		a, b := morfoio.Read2[int, int]()
		fmt.Println(squares(a, b))
	}
}

// T: O(1)
// S: O(1) extra space
func squares(startNum, endNum int) int {
	maxSquare := int(math.Floor(math.Sqrt(float64(endNum))))
	minSquare := int(math.Ceil(math.Sqrt(float64(startNum))))
	return maxSquare - minSquare + 1
}
